pennylane/optimize/gradient_descent.py

```diff
@@ -47,86 +48,110 @@ def update_stepsize(self, stepsize):
+                Must match shape of autograd derivative.
+            **kwargs : Variable length of keywords for the cost function
 
         Returns:
             array: the new variable values :math:`x^{(t+1)}`
         """
-        g, _ = self.compute_grad(objective_fn, x, grad_fn=grad_fn)
-        x_out = self.apply_grad(g, x)
 
-        return x_out
+        g, _ = self.compute_grad(objective_fn, args, kwargs, grad_fn=grad_fn)
+        new_args = self.apply_grad(g, args)
+
+        if len(new_args) == 1:
+            return new_args[0]
+
+        return new_args
 
     @staticmethod
-    def compute_grad(objective_fn, x, grad_fn=None):
+    def compute_grad(objective_fn, args, kwargs, grad_fn=None):
         r"""Compute gradient of the objective_fn at the point x and return it along with the
             objective function forward pass (if available).
 
         Args:
             objective_fn (function): the objective function for optimization
-            x (array): NumPy array containing the current values of the variables to be updated
+            args (tuple(array)): Tuple of NumPy arrays containing the current values for the
+                objection function
+            kwargs (dict): Keywords for the cost function
             grad_fn (function): Optional gradient function of the objective function with respect to
                 the variables ``x``. If ``None``, the gradient function is computed automatically.
+                Must match shape of autograd derivative.
 
         Returns:
             tuple: The NumPy array containing the gradient :math:`\nabla f(x^{(t)})` and the
                 objective function output. If ``grad_fn`` is provided, the objective function
                 will not be evaluted and instead ``None`` will be returned.
         """
         g = get_gradient(objective_fn) if grad_fn is None else grad_fn
-        grad = g(x)
+        grad = g(*args, **kwargs)
         forward = getattr(g, "forward", None)
 
         return grad, forward
 
-    def apply_grad(self, grad, x):
-        r"""Update the variables x to take a single optimization step. Flattens and unflattens
+    def apply_grad(self, grad, args):
+        r"""Update the variables to take a single optimization step. Flattens and unflattens
         the inputs to maintain nested iterables as the parameters of the optimization.
 
         Args:
             grad (array): The gradient of the objective
                 function at point :math:`x^{(t)}`: :math:`\nabla f(x^{(t)})`
-            x (array): the current value of the variables :math:`x^{(t)}`
+            x (tuple(array)): the current value of the variables :math:`x^{(t)}`
 
         Returns:
             array: the new values :math:`x^{(t+1)}`
         """
+        args_new = list(args)
+
+        trained_index = 0
+        for index, arg in enumerate(args):
+            if getattr(arg, "requires_grad", True):
```

Comment:
This is quite neat, the logic is almost identical to the logic on the left! The only difference being that we are now looping over all parameters, and only updating the trainable parameters.

Reply:
Less new things to figure out and debug

Reply:
@antalszava yes, this is for backwards compatibility. The initial versions of PL treated all python literals (such as floats) and sequences, as automatically differentiable. To preserve this behaviour, if the input does not have the requires grad attribute, we assume by default that it is differentiable to catch these cases.

At some point, we will deprecate this behaviour, and force users to pass trainable args as PL tensors with `requires_grad=True`. However, once we do this, we will need to update a lot of the demos/docstring examples!